### ISSUE TITLE
HMAC Signed JWT not verifying correctly

### ISSUE THREAD

**Issue by henyojess** [alvarosanchez/grails-spring-security-rest#172]
Changing the signature of an HMAC Signed JWT Bearer token still results in successful verification!

For example the following is a valid token:
`eyJhbGciOiJIUzI1NiJ9.eyJleHAiOjE0MjUwNDE2ODUsInN1YiI6Inh5eiIsInJvbGVzIjpbIlJPTEVfQVBJX1JFQUQiXSwiaWF0IjoxNDI1MDM4MDg1fQ.9TR4902AEhJ4XjoLBZhu34fXaAn2EY7yiVRR46OY1OM`

Replacing the signature with foobar still results in a valid token:
`eyJhbGciOiJIUzI1NiJ9.eyJleHAiOjE0MjUwNDE2ODUsInN1YiI6Inh5eiIsInJvbGVzIjpbIlJPTEVfQVBJX1JFQUQiXSwiaWF0IjoxNDI1MDM4MDg1fQ.foobar`

Am using
grails 2.4.4
spring-security-core:2.0-RC4 plugin
spring-security-rest:1.5.0.M2 plugin

from Config.groovy:
```
grails.plugin.springsecurity.filterChain.chainMap = [
    '/apiv1/**': 'JOINED_FILTERS,-exceptionTranslationFilter,-authenticationProcessingFilter,-securityContextPersistenceFilter,-rememberMeAuthenticationFilter',  // Stateless chain
    '/api/**': 'JOINED_FILTERS,-exceptionTranslationFilter,-authenticationProcessingFilter,-securityContextPersistenceFilter,-rememberMeAuthenticationFilter',  // Stateless chain
    '/**': 'JOINED_FILTERS,-restTokenValidationFilter,-restExceptionTranslationFilter'                                                                          // Traditional chain
]
grails.plugin.springsecurity.rest.token.storage.useJwt = true
grails.plugin.springsecurity.rest.token.storage.jwt.useSignedJwt = true
grails.plugin.springsecurity.rest.token.storage.jwt.secret = 'foofoofoofoofoofoobarbarbarbarbarbar'
grails.plugin.springsecurity.rest.token.storage.jwt.expiration = 3600
```
#### here is the debug output during successful authentication ####
```
2015-02-27 19:54:45,353 [http-bio-8080-exec-4] DEBUG matcher.AntPathRequestMatcher  - Checking match of request : '/api/login'; against '/apiv1/**'
2015-02-27 19:54:45,353 [http-bio-8080-exec-4] DEBUG matcher.AntPathRequestMatcher  - Checking match of request : '/api/login'; against '/api/**'
2015-02-27 19:54:45,354 [http-bio-8080-exec-4] DEBUG web.FilterChainProxy  - /api/login at position 1 of 7 in additional filter chain; firing Filter: 'MutableLogoutFilter'
2015-02-27 19:54:45,355 [http-bio-8080-exec-4] DEBUG web.FilterChainProxy  - /api/login at position 2 of 7 in additional filter chain; firing Filter: 'RestAuthenticationFilter'
2015-02-27 19:54:45,371 [http-bio-8080-exec-4] DEBUG rest.RestAuthenticationFilter  - Actual URI is /api/login; endpoint URL is /api/login
2015-02-27 19:54:45,371 [http-bio-8080-exec-4] DEBUG rest.RestAuthenticationFilter  - Applying authentication filter to this request
2015-02-27 19:54:45,409 [http-bio-8080-exec-4] DEBUG credentials.DefaultJsonPayloadCredentialsExtractor  - Extracted credentials from JSON payload. Username: xyz, password: [PROTECTED]
2015-02-27 19:54:45,420 [http-bio-8080-exec-4] DEBUG rest.RestAuthenticationFilter  - Trying to authenticate the request
2015-02-27 19:54:45,438 [http-bio-8080-exec-4] DEBUG authentication.ProviderManager  - Authentication attempt using org.springframework.security.authentication.dao.DaoAuthenticationProvider
2015-02-27 19:54:45,653 [http-bio-8080-exec-4] DEBUG hierarchicalroles.RoleHierarchyImpl  - getReachableGrantedAuthorities() - From the roles [ROLE_API_READ] one can reach [ROLE_API_READ] in zero or more steps.
2015-02-27 19:54:45,653 [http-bio-8080-exec-4] DEBUG rest.RestAuthenticationFilter  - Request authenticated. Storing the authentication result in the security context
2015-02-27 19:54:45,654 [http-bio-8080-exec-4] DEBUG rest.RestAuthenticationFilter  - Authentication result: org.springframework.security.authentication.UsernamePasswordAuthenticationToken@bf0f2d47: Principal: grails.plugin.springsecurity.userdetails.GrailsUser@1d199: Username: xyz; Password: [PROTECTED]; Enabled: true; AccountNonExpired: true; credentialsNonExpired: true; AccountNonLocked: true; Granted Authorities: ROLE_API_READ; Credentials: [PROTECTED]; Authenticated: true; Details: org.springframework.security.web.authentication.WebAuthenticationDetails@957e: RemoteIpAddress: 127.0.0.1; SessionId: null; Granted Authorities: ROLE_API_READ
2015-02-27 19:54:45,713 [http-bio-8080-exec-4] DEBUG jwt.AbstractJwtTokenGenerator  - Generated claim set: {"exp":1425041685,"sub":"xyz","roles":["ROLE_API_READ"],"iat":1425038085}
2015-02-27 19:54:45,813 [http-bio-8080-exec-4] DEBUG rest.RestAuthenticationFilter  - Generated token: eyJhbGciOiJIUzI1NiJ9.eyJleHAiOjE0MjUwNDE2ODUsInN1YiI6Inh5eiIsInJvbGVzIjpbIlJPTEVfQVBJX1JFQUQiXSwiaWF0IjoxNDI1MDM4MDg1fQ.9TR4902AEhJ4XjoLBZhu34fXaAn2EY7yiVRR46OY1OM
2015-02-27 19:54:45,815 [http-bio-8080-exec-4] DEBUG jwt.JwtTokenStorageService  - Nothing to store as this is a stateless implementation
2015-02-27 19:54:45,860 [http-bio-8080-exec-4] DEBUG rendering.DefaultRestAuthenticationTokenJsonRenderer  - Generated JSON:
{
   "username": "xyz",
   "roles": ["ROLE_API_READ"],
   "token_type": "Bearer",
   "access_token": "eyJhbGciOiJIUzI1NiJ9.eyJleHAiOjE0MjUwNDE2ODUsInN1YiI6Inh5eiIsInJvbGVzIjpbIlJPTEVfQVBJX1JFQUQiXSwiaWF0IjoxNDI1MDM4MDg1fQ.9TR4902AEhJ4XjoLBZhu34fXaAn2EY7yiVRR46OY1OM"
}
```

#### here is debug output while accessing a protected resource with the correct signature ####
```
2015-02-27 19:57:39,047 [http-bio-8080-exec-6] DEBUG matcher.AntPathRequestMatcher  - Checking match of request : '/apiv1'; against '/apiv1/**'
2015-02-27 19:57:39,048 [http-bio-8080-exec-6] DEBUG web.FilterChainProxy  - /apiV1 at position 1 of 7 in additional filter chain; firing Filter: 'MutableLogoutFilter'
2015-02-27 19:57:39,048 [http-bio-8080-exec-6] DEBUG web.FilterChainProxy  - /apiV1 at position 2 of 7 in additional filter chain; firing Filter: 'RestAuthenticationFilter'
2015-02-27 19:57:39,048 [http-bio-8080-exec-6] DEBUG rest.RestAuthenticationFilter  - Actual URI is /apiV1; endpoint URL is /api/login
2015-02-27 19:57:39,055 [http-bio-8080-exec-6] DEBUG web.FilterChainProxy  - /apiV1 at position 3 of 7 in additional filter chain; firing Filter: 'SecurityContextHolderAwareRequestFilter'
2015-02-27 19:57:39,057 [http-bio-8080-exec-6] DEBUG web.FilterChainProxy  - /apiV1 at position 4 of 7 in additional filter chain; firing Filter: 'GrailsAnonymousAuthenticationFilter'
2015-02-27 19:57:39,057 [http-bio-8080-exec-6] DEBUG web.FilterChainProxy  - /apiV1 at position 5 of 7 in additional filter chain; firing Filter: 'RestTokenValidationFilter'
2015-02-27 19:57:39,068 [http-bio-8080-exec-6] DEBUG bearer.BearerTokenReader  - Looking for bearer token in Authorization header, query string or Form-Encoded body parameter
2015-02-27 19:57:39,069 [http-bio-8080-exec-6] DEBUG bearer.BearerTokenReader  - Found bearer token in Authorization header
2015-02-27 19:57:39,069 [http-bio-8080-exec-6] DEBUG rest.RestTokenValidationFilter  - Token found: eyJhbGciOiJIUzI1NiJ9.eyJleHAiOjE0MjUwNDE2ODUsInN1YiI6Inh5eiIsInJvbGVzIjpbIlJPTEVfQVBJX1JFQUQiXSwiaWF0IjoxNDI1MDM4MDg1fQ.9TR4902AEhJ4XjoLBZhu34fXaAn2EY7yiVRR46OY1OM
2015-02-27 19:57:39,069 [http-bio-8080-exec-6] DEBUG rest.RestTokenValidationFilter  - Trying to authenticate the token
2015-02-27 19:57:39,071 [http-bio-8080-exec-6] DEBUG rest.RestAuthenticationProvider  - Trying to validate token eyJhbGciOiJIUzI1NiJ9.eyJleHAiOjE0MjUwNDE2ODUsInN1YiI6Inh5eiIsInJvbGVzIjpbIlJPTEVfQVBJX1JFQUQiXSwiaWF0IjoxNDI1MDM4MDg1fQ.9TR4902AEhJ4XjoLBZhu34fXaAn2EY7yiVRR46OY1OM
2015-02-27 19:57:39,082 [http-bio-8080-exec-6] DEBUG jwt.JwtTokenStorageService  - Parsed an HMAC signed JWT
2015-02-27 19:57:39,096 [http-bio-8080-exec-6] DEBUG jwt.JwtTokenStorageService  - Successfully verified JWT
2015-02-27 19:57:39,098 [http-bio-8080-exec-6] DEBUG rest.RestAuthenticationProvider  - Authentication result: com.odobo.grails.plugin.springsecurity.rest.RestAuthenticationToken@bf0e92f9: Principal: org.springframework.security.core.userdetails.User@1d199: Username: xyz; Password: [PROTECTED]; Enabled: true; AccountNonExpired: true; credentialsNonExpired: true; AccountNonLocked: true; Granted Authorities: ROLE_API_READ; Credentials: [PROTECTED]; Authenticated: true; Details: null; Granted Authorities: ROLE_API_READ
2015-02-27 19:57:39,098 [http-bio-8080-exec-6] DEBUG rest.RestTokenValidationFilter  - Token authenticated. Storing the authentication result in the security context
2015-02-27 19:57:39,098 [http-bio-8080-exec-6] DEBUG rest.RestTokenValidationFilter  - Authentication result: com.odobo.grails.plugin.springsecurity.rest.RestAuthenticationToken@bf0e92f9: Principal: org.springframework.security.core.userdetails.User@1d199: Username: xyz; Password: [PROTECTED]; Enabled: true; AccountNonExpired: true; credentialsNonExpired: true; AccountNonLocked: true; Granted Authorities: ROLE_API_READ; Credentials: [PROTECTED]; Authenticated: true; Details: null; Granted Authorities: ROLE_API_READ
2015-02-27 19:57:39,100 [http-bio-8080-exec-6] DEBUG rest.RestTokenValidationFilter  - Continuing the filter chain
2015-02-27 19:57:39,100 [http-bio-8080-exec-6] DEBUG web.FilterChainProxy  - /apiV1 at position 6 of 7 in additional filter chain; firing Filter: 'ExceptionTranslationFilter'
2015-02-27 19:57:39,100 [http-bio-8080-exec-6] DEBUG web.FilterChainProxy  - /apiV1 at position 7 of 7 in additional filter chain; firing Filter: 'FilterSecurityInterceptor'
2015-02-27 19:57:39,190 [http-bio-8080-exec-6] DEBUG intercept.FilterSecurityInterceptor  - Secure object: FilterInvocation: URL: /apiV1; Attributes: [ROLE_API_READ]
2015-02-27 19:57:39,191 [http-bio-8080-exec-6] DEBUG intercept.FilterSecurityInterceptor  - Previously Authenticated: com.odobo.grails.plugin.springsecurity.rest.RestAuthenticationToken@bf0e92f9: Principal: org.springframework.security.core.userdetails.User@1d199: Username: xyz; Password: [PROTECTED]; Enabled: true; AccountNonExpired: true; credentialsNonExpired: true; AccountNonLocked: true; Granted Authorities: ROLE_API_READ; Credentials: [PROTECTED]; Authenticated: true; Details: null; Granted Authorities: ROLE_API_READ
2015-02-27 19:57:39,191 [http-bio-8080-exec-6] DEBUG hierarchicalroles.RoleHierarchyImpl  - getReachableGrantedAuthorities() - From the roles [ROLE_API_READ] one can reach [ROLE_API_READ] in zero or more steps.
2015-02-27 19:57:39,192 [http-bio-8080-exec-6] DEBUG intercept.FilterSecurityInterceptor  - Authorization successful
2015-02-27 19:57:39,192 [http-bio-8080-exec-6] DEBUG intercept.FilterSecurityInterceptor  - RunAsManager did not change Authentication object
2015-02-27 19:57:39,192 [http-bio-8080-exec-6] DEBUG web.FilterChainProxy  - /apiV1 reached end of additional filter chain; proceeding with original chain
2015-02-27 19:57:39,236 [http-bio-8080-exec-6] DEBUG access.ExceptionTranslationFilter  - Chain processed normally
```

#### here is debug output while accessing a protected resource with the INCORRECT signature i.e. replaced with foobar ####

```
2015-02-27 19:58:36,543 [http-bio-8080-exec-8] DEBUG matcher.AntPathRequestMatcher  - Checking match of request : '/apiv1'; against '/apiv1/**'
2015-02-27 19:58:36,543 [http-bio-8080-exec-8] DEBUG web.FilterChainProxy  - /apiV1 at position 1 of 7 in additional filter chain; firing Filter: 'MutableLogoutFilter'
2015-02-27 19:58:36,543 [http-bio-8080-exec-8] DEBUG web.FilterChainProxy  - /apiV1 at position 2 of 7 in additional filter chain; firing Filter: 'RestAuthenticationFilter'
2015-02-27 19:58:36,543 [http-bio-8080-exec-8] DEBUG rest.RestAuthenticationFilter  - Actual URI is /apiV1; endpoint URL is /api/login
2015-02-27 19:58:36,544 [http-bio-8080-exec-8] DEBUG web.FilterChainProxy  - /apiV1 at position 3 of 7 in additional filter chain; firing Filter: 'SecurityContextHolderAwareRequestFilter'
2015-02-27 19:58:36,544 [http-bio-8080-exec-8] DEBUG web.FilterChainProxy  - /apiV1 at position 4 of 7 in additional filter chain; firing Filter: 'GrailsAnonymousAuthenticationFilter'
2015-02-27 19:58:36,544 [http-bio-8080-exec-8] DEBUG web.FilterChainProxy  - /apiV1 at position 5 of 7 in additional filter chain; firing Filter: 'RestTokenValidationFilter'
2015-02-27 19:58:36,544 [http-bio-8080-exec-8] DEBUG bearer.BearerTokenReader  - Looking for bearer token in Authorization header, query string or Form-Encoded body parameter
2015-02-27 19:58:36,545 [http-bio-8080-exec-8] DEBUG bearer.BearerTokenReader  - Found bearer token in Authorization header
2015-02-27 19:58:36,545 [http-bio-8080-exec-8] DEBUG rest.RestTokenValidationFilter  - Token found: eyJhbGciOiJIUzI1NiJ9.eyJleHAiOjE0MjUwNDE2ODUsInN1YiI6Inh5eiIsInJvbGVzIjpbIlJPTEVfQVBJX1JFQUQiXSwiaWF0IjoxNDI1MDM4MDg1fQ.foobar
2015-02-27 19:58:36,545 [http-bio-8080-exec-8] DEBUG rest.RestTokenValidationFilter  - Trying to authenticate the token
2015-02-27 19:58:36,545 [http-bio-8080-exec-8] DEBUG rest.RestAuthenticationProvider  - Trying to validate token eyJhbGciOiJIUzI1NiJ9.eyJleHAiOjE0MjUwNDE2ODUsInN1YiI6Inh5eiIsInJvbGVzIjpbIlJPTEVfQVBJX1JFQUQiXSwiaWF0IjoxNDI1MDM4MDg1fQ.foobar
2015-02-27 19:58:36,546 [http-bio-8080-exec-8] DEBUG jwt.JwtTokenStorageService  - Parsed an HMAC signed JWT
2015-02-27 19:58:36,548 [http-bio-8080-exec-8] DEBUG jwt.JwtTokenStorageService  - Successfully verified JWT
2015-02-27 19:58:36,548 [http-bio-8080-exec-8] DEBUG rest.RestAuthenticationProvider  - Authentication result: com.odobo.grails.plugin.springsecurity.rest.RestAuthenticationToken@bf0e92f9: Principal: org.springframework.security.core.userdetails.User@1d199: Username: xyz; Password: [PROTECTED]; Enabled: true; AccountNonExpired: true; credentialsNonExpired: true; AccountNonLocked: true; Granted Authorities: ROLE_API_READ; Credentials: [PROTECTED]; Authenticated: true; Details: null; Granted Authorities: ROLE_API_READ
2015-02-27 19:58:36,548 [http-bio-8080-exec-8] DEBUG rest.RestTokenValidationFilter  - Token authenticated. Storing the authentication result in the security context
2015-02-27 19:58:36,548 [http-bio-8080-exec-8] DEBUG rest.RestTokenValidationFilter  - Authentication result: com.odobo.grails.plugin.springsecurity.rest.RestAuthenticationToken@bf0e92f9: Principal: org.springframework.security.core.userdetails.User@1d199: Username: xyz; Password: [PROTECTED]; Enabled: true; AccountNonExpired: true; credentialsNonExpired: true; AccountNonLocked: true; Granted Authorities: ROLE_API_READ; Credentials: [PROTECTED]; Authenticated: true; Details: null; Granted Authorities: ROLE_API_READ
2015-02-27 19:58:36,549 [http-bio-8080-exec-8] DEBUG rest.RestTokenValidationFilter  - Continuing the filter chain
2015-02-27 19:58:36,549 [http-bio-8080-exec-8] DEBUG web.FilterChainProxy  - /apiV1 at position 6 of 7 in additional filter chain; firing Filter: 'ExceptionTranslationFilter'
2015-02-27 19:58:36,549 [http-bio-8080-exec-8] DEBUG web.FilterChainProxy  - /apiV1 at position 7 of 7 in additional filter chain; firing Filter: 'FilterSecurityInterceptor'
2015-02-27 19:58:36,550 [http-bio-8080-exec-8] DEBUG intercept.FilterSecurityInterceptor  - Secure object: FilterInvocation: URL: /apiV1; Attributes: [ROLE_API_READ]
2015-02-27 19:58:36,550 [http-bio-8080-exec-8] DEBUG intercept.FilterSecurityInterceptor  - Previously Authenticated: com.odobo.grails.plugin.springsecurity.rest.RestAuthenticationToken@bf0e92f9: Principal: org.springframework.security.core.userdetails.User@1d199: Username: xyz; Password: [PROTECTED]; Enabled: true; AccountNonExpired: true; credentialsNonExpired: true; AccountNonLocked: true; Granted Authorities: ROLE_API_READ; Credentials: [PROTECTED]; Authenticated: true; Details: null; Granted Authorities: ROLE_API_READ
2015-02-27 19:58:36,550 [http-bio-8080-exec-8] DEBUG hierarchicalroles.RoleHierarchyImpl  - getReachableGrantedAuthorities() - From the roles [ROLE_API_READ] one can reach [ROLE_API_READ] in zero or more steps.
2015-02-27 19:58:36,551 [http-bio-8080-exec-8] DEBUG intercept.FilterSecurityInterceptor  - Authorization successful
2015-02-27 19:58:36,551 [http-bio-8080-exec-8] DEBUG intercept.FilterSecurityInterceptor  - RunAsManager did not change Authentication object
2015-02-27 19:58:36,551 [http-bio-8080-exec-8] DEBUG web.FilterChainProxy  - /apiV1 reached end of additional filter chain; proceeding with original chain
2015-02-27 19:58:36,554 [http-bio-8080-exec-8] DEBUG access.ExceptionTranslationFilter  - Chain processed normally
```

**Comment by alvarosanchez** [comment]
Thanks for reporting this. I will create a test to reproduce it and then make sure it's fixed.

For the verification, I'm using an external library. I will double check how I'm using it as I might be missing something.

**Comment by henyojess** [comment]
Tried using Encrypted JWT; there were no problems i.e. the expected behavior was observed

**Comment by marcos-carceles** [comment]
+1. I identified this in our project too.

**Comment by alvarosanchez** [comment]
Fixed by @marcos-carceles in #177
